src/Authentication/PasswordAuthentication.php

```
@@ -33,66 +49,91 @@ class PasswordAuthentication implements AuthenticationInterface
     }
 
     /**
-     * @inheritDoc
-     *
+     * @param RequestInterface $request
+     * @return RequestInterface
      * @throws AuthenticationException
+     * @throws ClientExceptionInterface
      */
     public function authenticate(RequestInterface $request)
     {
-        if (is_null($this->cookie)) {
-            $this->cookie = $this->retrieveCookie();
+        if (is_null($this->authenticationCookies)) {
+            $this->authenticationCookies = $this->retrieveCookies();
         }
 
-        return $this->cookie->withCookieHeader($request);
+        return $request->withHeader('Cookie', implode('; ', $this->authenticationCookies));
     }
 
     /**
-     * @return CookieJarInterface
+     * @return string[]
      * @throws AuthenticationException
+     * @throws ClientExceptionInterface
      */
-    protected function retrieveCookie()
+    protected function retrieveCookies()
     {
-        $client = new GuzzleHttpClient(['cookies' => true]);
-
-        $jar = new CookieJar();
-
-        $response = $client->post(
-            $this->host . '/j_security_check',
-            [
-                'cookies' => $jar,
-
-                'form_params' => [
+        $request = $this->requestFactory
+            ->createRequest('POST', $this->host . '/j_security_check')
+            ->withHeader('Content-Type', 'application/x-www-form-urlencoded')
+            ->withBody(
+                $this->streamFactory->createStream(http_build_query([
                     'j_username' => $this->username,
                     'j_password' => $this->password
-                ]
-            ]
-        );
+                ]))
+            );
 
-        if ($response->getStatusCode() === 200) {
-            return $jar;
+        $response = $this->httpClient->sendRequest($request);
+
+        if ($response->getStatusCode() < 400 && $response->hasHeader('Set-Cookie')) {
+            return $this->extractCookies($response);
         } else {
```

Else is unneeded here
You could throw an Exception without it

---

Good hint, will optimize that